Enhanced EOL support

MysterAitch opened the pull request:
Fixes #2647 

This PR adds an enum to represent different types of end-of-line / line-separator characters. 

- It also adds a utility method which allows you to "detect" which style of line-separator a given string uses.

- This is incorporated into the parsing process by reading in the content of the given provider and adding this as a data field within the top-most node. 

- This is then used as part of the CSM and pretty printer where new added nodes will reflect this (defaulting to the system's eol character if it is unknown/ambiguous e.g. it is a single line or there are mixed endings).

Reading in the provider content and then creating a brand new `StringProvider` using this string is a bit "hacky" and I'm sure there must be a better way to do it, I just don't know what it is -- comments and suggestions welcomed!

MysterAitch committed:
added enum to represent different styles of line endings, and marked `Utils.EOL` as deprecated

MysterAitch committed:
Added test case, comments, and fixed a few bugs...

MysterAitch committed:
fixes #2647 -- nodes now have a data item with detail about which line ending was found during parsing -- this is then used during the lexical preservation csmtoken insertion/diffing to ensure that the "new" newlines are of the same type as the parent....

MysterAitch committed:
added helper method and checks where used to default to system's line separator..

MysterAitch committed:
extracted out method which gets a string from a provider

MysterAitch committed:
added method that allows passing a default line ending if there either one isn't set or it is a meta-value e.g. EMPTY or NONE

MysterAitch committed:
bug fix where the node has no EOL chars already -- default to using system's line ending

MysterAitch committed:
fixed the CsmToken based only on a given token type int -- the content should be unescaped line separators, but the lookup returns escaped line separators; deprecated isSpaceOrTab and added isWhitespaceButNotEndOfLine as a replacement;

MysterAitch committed:
added comments to the variety of whitespace characters treated as a space within the grammar (e.g. non-breaking spaces and tab characters)

MysterAitch commented on javaparser-core/src/main/java/com/github/javaparser/JavaParser.java:
```diff
@@ -114,6 +132,18 @@ private GeneratedJavaParser getParserForProvider(Provider provider) {
+        String x = providerToString(provider);
+        // close the old provider then open a new one...
+        try {
+            provider.close();
```
Surely there's a better way to achieve this than consuming the whole stream, interpreting it, and then forwarding a brand new stream (`provider`) based on this new string... 

Note that the original stream will be unavailable due to having been fully consumed.

matozoid replied:
I think you want to create a preprocessor for this, like the unicode escaping preprocessor. Then add a configuration option so those who don't care about this feature can turn it off :-)

MysterAitch replied:
I didn't really understand the unicode preprocessor(!)

What I could determine is that it modifies the stream, but isn't able to store any state or pass any results on..? Or maybe I misunderstood.

The other aspect (I can't remember if I mentioned this in another comment) is that during the pre-processing stage the processors don't seem to have access to either the config or the the resulting nodes.

Presumably it's inappropriate to store this on the processor itself... 

If that's something you can give guidance on, then that would be great! Otherwise I'll ruminate and think on this some more :)

MysterAitch replied:
I've done removed this and replaced it with a preprocessor - marking as "resolved".

MysterAitch commented on javaparser-core/src/main/java/com/github/javaparser/JavaParser.java:
```diff
@@ -123,6 +153,10 @@ private GeneratedJavaParser getParserForProvider(Provider provider) {
             N resultNode = start.parse(parser);
```
Note that at this point, the line endings have already defaulted to the system's EOL, meaning that post processors don't seem to have access to the "original" underlying line separators.

MysterAitch commented on javaparser-core/src/main/java/com/github/javaparser/printer/concretesyntaxmodel/CsmElement.java:
```diff
@@ -99,7 +99,11 @@ static CsmElement semicolon() {
+    }
+
+    static CsmElement newline(LineEnding lineEnding) {
+        return new CsmToken(eolTokenKind(lineEnding), lineEnding.toString());
```
Not passing in a parameter here defaults to the system's EOL.

MysterAitch commented on javaparser-core-testing/src/test/java/com/github/javaparser/utils/TestUtils.java:
```diff
@@ -194,13 +194,13 @@ public static void assertExpressionValid(String expression) {
      */
     public static void assertEqualsNoEol(String expected, String actual) {
-        assertEquals(normalizeEolInTextBlock(expected, EOL), actual);
+        assertEquals(normalizeEolInTextBlock(expected, EOL), normalizeEolInTextBlock(actual, EOL));
```
My understanding/interpretation of the method name `assertEqualsNoEol` is that it should assert that everything except EOL is equal. Given this, my inclination is that we should normalise to `LineSeparator.ARBITRARY` to explicitly indicate this. 

Note that this would contradict the javadoc which **ALSO** specifies ".. and that any EOL characters in "actual" are correct for the platform.". My preference here would be to have a separate assertion for this.

matozoid replied:
It is there to compare a string with "\n"'s in it to some output which contains some kind of line ending, which is normalized to "\n" so they match. If I remember correctly :-)

MysterAitch replied:
For the purpose of this assert (and how I've ever used it based on the name), the important bit is that they're the **same** line endings as opposed to being any particular line ending.

The javadoc suggests otherwise though -- that it should match the system's line ending.

matozoid replied:
Ah yes, the other way around. Well, the important thing is that it is really nice to be able to say "\n" in an expectation and not have to bother with Windows/Unix differences. Yeah, I see what you're doing now. Okay.

MysterAitch commented on javaparser-core/src/main/java/com/github/javaparser/JavaParser.java:
```diff
@@ -100,6 +101,23 @@ private GeneratedJavaParser getParserForProvider(Provider provider) {
     }
 
+    private String providerToString(Provider provider) {
+        // FIXME: Surely there's a better way...?
```
Commenting this line to flag this up for attention -- this method consumes the whole provider stream and returns a string.

MysterAitch replied:
I've done removed this and replaced it with a preprocessor - marking as "resolved".

MysterAitch commented on javaparser-core/src/main/java/com/github/javaparser/ast/Node.java:
```diff
@@ -722,6 +757,9 @@ protected SymbolResolver getSymbolResolver() {
     public static final DataKey<SymbolResolver> SYMBOL_RESOLVER_KEY = new DataKey<SymbolResolver>() {
     };
 
+    public static final DataKey<LineEnding> LINE_ENDING_KEY = new DataKey<LineEnding>() {
```
Just noticed this. It might be able to be made private -- originally there wasn't an accessor method on Node. Maybe we can also add in a mutator / setter method.

MysterAitch replied:
resolved -- leaving as public

matozoid commented on javaparser-core/src/main/java/com/github/javaparser/TokenTypes.java:
```diff
@@ -51,19 +66,23 @@ public static boolean isComment(int kind) {
         }
-        if (EOL.equals("\r")) {
+        if (lineEnding.equalsString(LineEnding.CR)) {
             return OLD_MAC_EOL;
```
Would it be a bad idea to put the corresponding tokens in the enum too?

MysterAitch replied:
I hesitated there because `WINDOWS_EOL` AND `OLD_MAC_EOL` refer to static ints within the grammar. 

The other aspect that caused hesitation there is to keep the enum relatively "pure" and not needing to (necessarily) ever be aware of the grammar stuff.

Neither are show-stoppers though.

MysterAitch replied:
Unless, of course, I'm misunderstanding you?

matozoid replied:
Nah, I think you understand right and it's a fine decision. It made me think of this thing by the way: https://github.com/javaparser/javaparser/blob/master/javaparser-core/src/main/java/com/github/javaparser/JavaToken.java - it is my attempt to define the definitive "token" class, so that all other weird int-tokens and methods operating on them could be removed. Nothing relevant now, just info.

MysterAitch replied:
Plan -- leaving as-is for now.. it can be improved in a future release.

MysterAitch commented on javaparser-core/src/main/java/com/github/javaparser/printer/concretesyntaxmodel/CsmTextBlock.java:
```diff
@@ -34,7 +35,8 @@ public CsmTextBlock(ObservableProperty property) {
     public void prettyPrint(Node node, SourcePrinter printer) {
-        printer.print("\"\"\"\n");
+        LineEnding lineEnding = node.getLineEndingStyleOrDefault(LineEnding.SYSTEM);
+        printer.print("\"\"\"" + lineEnding.toString());
```
As highlighted by @matozoid , textblocks are picky about line endings 

https://github.com/javaparser/javaparser/pull/2685#issuecomment-631008605

https://openjdk.java.net/jeps/378#1--Line-terminators

MysterAitch replied:
Hmm, It has just occurred to me (in the couple mins since posting my reply) that maybe the **VALUE** is always interpreted as \n, but it is unclear to me whether the **source code** within a java file can have different line endings.

Regardless, it needs some more attention +- a comment to explain.

matozoid replied:
That should definitely remain "\n", textblocks never generate "\r\n" or so.

MysterAitch replied:
Ah yes, just realised that at this point we're it the CSM dealing with interpreted values (as opposed to source code) so my last comment isn't applicable.

matozoid replied:
And the change we're discussing breaks text blocks, because now they generate platform specific EOL's, which they should not do :-)

MysterAitch replied:
Reverted this and added a line comment to explain

MysterAitch committed:
Revert change to CsmTextBlock and added documentation

MysterAitch committed:
Merge branch 'master' into issue_2647_eol

MysterAitch committed:
tweak test to give better failure messages

MysterAitch committed:
Merge branch 'issue_2647_eol' of https://github.com/MysterAitch/javaparser into issue_2647_eol

MysterAitch committed:
Added more descriptive assertion message, and javadoc

MysterAitch committed:
.. now the output actually prints the output in human-readable form.. oops!

MysterAitch committed:
attempt two at human-readable output

MysterAitch committed:
include the system's EOL in the test output

MysterAitch committed:
edited TestUtils#readResource to read the file "as-is" (i.e. character by character). If it is important for you to define the line endings used, provide a LineEnding as a second parameter else use TestUtils#readResourceUsingSystemEol

MysterAitch committed:
editing another test case which presumes system EOL -- this commit provides more descriptive output from the test's utility method

MysterAitch committed:
Introduced TestUtils.assertEqualsString and TestUtils.assertEqualsStringIgnoringEol, and edited some tests to this.

MysterAitch committed:
converted usages of `assertEqualsNoEol` to `assertEqualsStringIgnoringEol`

MysterAitch committed:
updated appveyor.yml to set autocrlf to true -- appveyor defaults to `input` meaning that it checks out the code using the repo's line endings not the system's line endings

this was discovered / came up as a problem in #2685 where javaparser started to respect the line endings of test resources instead of reading line-by-line and inserting the system's line endings

MysterAitch committed:
Merge branch 'master' into issue_2647_eol

MysterAitch committed:
Merge branch 'master' into issue_2647_eol

MysterAitch committed:
Merge branch 'master' into issue_2647_eol

MysterAitch committed:
initial switch to using line ending detection as a preprocessor

MysterAitch committed:
renamed uses of `toString()` to `toRawString()` to emphasise that the caller wants the actual line separator "as-is" and without any additional escaping; added human-readable "descriptions" to the LineEnding enums; progress on LineEndingProcessingProvider.java -- now functional and tests passing;

MysterAitch committed:
renamed LineEnding to LineSeparator

MysterAitch committed:
renamed assertEqualsNoEol to assertEqualsStringIgnoringEol

MysterAitch committed:
renamed EOL to SYSTEM_EOL to make it explicit that it refers to the host's line separator

MysterAitch committed:
renamed variable/config option re: detecting line endings as opposed to retaining them

MysterAitch committed:
renamed variable/config option re: detecting line endings as opposed to retaining them

MysterAitch committed:
renamed toRawString and toEscapedString to asRawString and asEscapedString

MysterAitch committed:
renamed variable/config option re: detecting line endings as opposed to retaining them

MysterAitch committed:
typo fix in tests -- accidentally comparing string to enum LineSeparator

MysterAitch committed:
updated changelog.md
